src/pk_encryption_circuit.rs

```diff
@@ -0,0 +1,714 @@
+            let rhs = gate.mul_add(ctx_gate, k1_at_gamma, k0i_constants[z], pk0_u);
+
+            // rhs = rhs + r1i(gamma) * qi
+            let rhs = gate.mul_add(ctx_gate, r1i_at_gamma, qi_constants[z], rhs);
+
+            // rhs = rhs + r2i(gamma) * cyclo(gamma)
+            let rhs = gate.mul_add(ctx_gate, r2i_at_gamma, cyclo_at_gamma_assigned, rhs);
+            let lhs = ct0is_assigned[z].enforce_eval_at_gamma(ctx_rlc, rlc);
+
+            // LHS(gamma) = RHS(gamma)
+            let res = gate.is_equal(ctx_gate, lhs, rhs);
+            gate.assert_is_const(ctx_gate, &res, &F::from(1));
+        }
+
+        for z in 0..ct1is_assigned.len() {
+            let pk1_u = gate.mul_add(ctx_gate, pk1_qi_at_gamma[z], u_at_gamma, e1_at_gamma);
+
+            let p1i_at_gamma = p1is_assigned[z].enforce_eval_at_gamma(ctx_rlc, rlc);
+            let p2i_at_gamma = p2is_assigned[z].enforce_eval_at_gamma(ctx_rlc, rlc);
+
+            //rhs = pk1_u + p2i * cyclo(gamma)
+            let rhs = gate.mul_add(ctx_gate, p2i_at_gamma, cyclo_at_gamma_assigned, pk1_u);
+
+            //rhs = rhs + p1s * qi
+            let rhs = gate.mul_add(ctx_gate, p1i_at_gamma, qi_constants[z], rhs);
+
+            let lhs = ct1is_assigned[z].enforce_eval_at_gamma(ctx_rlc, rlc);
+
+            let res = gate.is_equal(ctx_gate, lhs, rhs);
+            gate.assert_is_const(ctx_gate, &res, &F::from(1));
+        }
+    }
+
+    fn instances(&self) -> Vec<Vec<F>> {
+        let mut instance = vec![];
+        for pk0 in self.pk0_qi.iter() {
+            let pk0_poly = Poly::<F>::new(pk0.clone());
+            instance.push(pk0_poly.coefficients[0]);
+        }
+        for pk1 in self.pk1_qi.iter() {
+            let pk1_poly = Poly::<F>::new(pk1.clone());
+            instance.push(pk1_poly.coefficients[0]);
+        }
+        for ct0i in self.ct0is.iter() {
+            let ct0i_poly = Poly::<F>::new(ct0i.clone());
+            instance.push(ct0i_poly.coefficients[0]);
+        }
+        for ct1i in self.ct1is.iter() {
+            let ct1i_poly = Poly::<F>::new(ct1i.clone());
+            instance.push(ct1i_poly.coefficients[0]);
+        }
+        vec![instance]
+    }
+}
+#[cfg(test)]
+mod test {
+    use std::{fs::File, io::Read};
+
+    use axiom_eth::{
+        halo2_proofs::{
+            dev::MockProver,
+            plonk::{keygen_pk, keygen_vk},
+        },
+        halo2curves::bn256::Fr,
+        rlc::{
+            circuit::{builder::RlcCircuitBuilder, instructions::RlcCircuitInstructions},
+            utils::executor::RlcExecutor,
+        },
+    };
+    use halo2_base::{
+        gates::circuit::CircuitBuilderStage,
+        utils::{
+            fs::gen_srs,
+            testing::{check_proof_with_instances, gen_proof_with_instances},
+        },
+    };
+
+    use crate::constants::pk_enc_constants::pk_enc_constants_1024_15x60_65537::R1_BOUNDS;
+
+    use super::{test_params, BfvPkEncryptionCircuit};
+
+    #[test]
+    fn test_pk_enc_valid() {
```

Comment:
This test should use the mock prover. Otherwise it's same as the next test `pk_enc_full_prover`